### ISSUE TITLE
should `sourmash sketch` fail on empty input files?

### ISSUE THREAD

**Issue by ctb** [sourmash-bio/sourmash#1887]
So @bluegenes was looking at some new code for `sourmash sketch` in #1885 and [noticed that we skip over empty sequence files with just a warning](https://github.com/sourmash-bio/sourmash/pull/1885#discussion_r827488718). She made the excellent point that this isn't good behavior in a lot of situations:
```
        # now, set up to iterate over sequences.
        with screed.open(filename) as screed_iter:
            if not screed_iter:
                notify(f"no sequences found in '{filename}'?!")
```

I got curious and dug into the tests and realized that all three of the tests for this (all named `_noinput`) were for stdin input.

I think we should change this code so that `sourmash sketch` fails when run on empty files, maybe with a special case for stdin.

An alternative would be to have a pretty explicit report-out at the end saying "<this many> empty files."

thoughts?

**Comment by bluegenes** [comment]
My thinking:

I don't want a single empty input file to stop sketching of independent files (e.g. during bulk/batch sketching), but I also don't want to ignore empties/ make databases with empty sketches made from empty input files.

> An alternative would be to have a pretty explicit report-out at the end saying " empty files."

this seems like it would work? Esp as it should contain enough information to allow easy re-download/re-sketch of the empty files.

I'm also happy with sketching failing as the default, as long as there is a way we can force it to temporarily ignore & report empty files...

OR, perhaps better: we could enable a relatively quick `sourmash sketch fromfile --check-inputs` or similar that allows you to check inputfiles FIRST, before you set a large job running, so problems can be solved ahead of time? Then fail by default, but allow forcing through w/reporting?

**Comment by taylorreiter** [comment]
I had requested this "feature" because of IBD and snakemake. I had a lot of files that were empty, so I just wanted an empty sig out the back end so my pipeline wouldn't fail for a bunch of files. So if the default behavior is to not do this, can we have a force flag or something to create an empty sig or something? it was really really handy

**Comment by ctb** [comment]
thanks - excellent feedback.

I, too, am finding it annoying to have things crash on empty files. I'll work it into the overall revisions. I would note here that I anticipate very different use cases and user bases for `sketch fromfile` than `sketch dna|protein|translate` so we might also make that distinction.

And, regardless, an `--empty-is-ok` flag or something can happen!

**Comment by ctb** [comment]
hah, just ran across this issue again after hand-waving away [this comment](https://github.com/sourmash-bio/sourmash/pull/1885#discussion_r840843520) --

[bluegenes](https://github.com/bluegenes) [19 hours ago](https://github.com/sourmash-bio/sourmash/pull/1885#discussion_r840843520)

>looks like you can't force your way through empty files here (which is good, but I seem to remember an issue about this :). So the workflow, if I have it right, is to check the csv first, fix/remove empty files, then actually build?

@[ctb](https://github.com/ctb) ctb [19 hours ago](https://github.com/sourmash-bio/sourmash/pull/1885#discussion_r840881915)
>yes, I think that for batch construction it doesn't make sense. You can force your way through missing files but not empty files.

so let me explain myself.

As tessa observed, in the new `sketch fromfile` command to do bulk sketching, you cannot ignore empty files. That is, if you put a filename in the CSV `fromfile`, and it is an empty file, `sourmash sketch fromfile` **will** barf.

you _can_ optionally ignore _missing_ files - that is, if you ask for a protein sketch and there is an empty cell in the protein_filename column, you can specify `--ignore-missing`.

This is because when I was playing around with `sketch fromfile`, I realized that if you were building `fromfile` files to bulk-construct databases, it was really up to you whether you included empty files in your `fromfile` CSV spreadsheet, or simply made them blank cells, and it seemed way less hacky to say:

* if you have an empty file, it's probably a mistake
* on the off chance that the file is intentionally empty, you can always simply omit it from your spreadsheet during the process of constructing the spreadsheet, and then specify `--ignore-missing` to `sourmash sketch fromfile`.

This different from the situation where you are bulk sketching on the command line with `sketch dna|protein|translate`, where I think you may have less control over what's in your directory.

tl;dr I think `--ignore-empty` belongs in the preparation scripts in https://github.com/sourmash-bio/database-examples and (as it currently is) in the `sketch dna|protein|translate` commands, but probably not on `sketch fromfile` where you have a lot more control.

**Comment by ctb** [comment]
> OR, perhaps better: we could enable a relatively quick `sourmash sketch fromfile --check-inputs` or similar that allows you to check inputfiles FIRST, before you set a large job running, so problems can be solved ahead of time? Then fail by default, but allow forcing through w/reporting?

I think this is probably best done during the fromfile preparation stage.

**Comment by ctb** [comment]
After re-reading this issue, Imma gonna close it, and if after some more real world experiences people hate the way things currently work, we can revisit :).
